File: python/day16.py
import sys
import heapq
from enum import Enum
from pathlib import Path
from collections import deque

PRINT = False

START = "S"
EMPTY = "."
WALL = "#"
END = "E"

NORTH, EAST, SOUTH, WEST = range(4)
class Direction(Enum):
    NORTH = 0
    EAST = 1
    SOUTH = 2
    WEST = 3
    
    @staticmethod
    def to_str(direction):
        match direction:
            case Direction.NORTH:
                return "^"
            case Direction.EAST:
                return ">"
            case Direction.SOUTH:
                return "v"
            case Direction.WEST:
                return "<"
        return "X"
    
ROTATIONS = ((-1, 0), (0, 1), (1, 0), (0, -1))

def show_history(grid, history, show_path=True):
    if not PRINT:
        return
    grid = grid.copy()
    history_dict = {}
    for r, c, direction in history:
        history_dict[(r, c)] = direction
    for r, row in enumerate(grid):
        for c, elem in enumerate(row):
            if (r, c) in history_dict:
                if show_path:
                    grid[r][c] = Direction.to_str(Direction(direction))
                else:
                    grid[r][c] = "O"
            
    for row in grid:
        print("".join(row))
    print()

def part1():
    
    grid = parse()
    ROWS, COLS = len(grid), len(grid[0])
    
    initial_pos = (-1, -1)
    for r in range(ROWS):
        for c in range(COLS):
            if grid[r][c] == START:
                initial_pos = (r, c)
                
    min_heap = [(0, initial_pos[0], initial_pos[1], EAST, [])] # (score, r, c, direction, history)
    visited = set()
    while min_heap:
        score, r, c, direction, history = heapq.heappop(min_heap)
        dr, dc = ROTATIONS[direction]
        if (
            r not in range(ROWS) or
            c not in range(COLS) or
            grid[r][c] == WALL or
            (r, c, dr, dc) in visited
        ):
            continue
        if grid[r][c] == END:
            break
        visited.add((r, c, dr, dc))
        
        # go straight
        heapq.heappush(min_heap, (score + 1, r + dr, c + dc, direction, history + [(r, c, direction)]))
        
        # try turning
        heapq.heappush(min_heap, (score + 1000, r, c, (direction + 1) % 4, history + [(r, c, direction)]))
        heapq.heappush(min_heap, (score + 1000, r, c, (direction - 1) % 4, history + [(r, c, direction)]))
    
    
    show_history(grid, history)
    return score

def part2():
    grid = parse()
    ROWS, COLS = len(grid), len(grid[0])
    
    initial_pos = (-1, -1)
    for r in range(ROWS):
        for c in range(COLS):
            if grid[r][c] == START:
                initial_pos = (r, c)
                break
        if initial_pos != (-1, -1):
            break
    
    sr, sc = initial_pos
    
    pq = [(0, sr, sc, 0, 1)]  # (cost, r, c, dr, dc)
    lowest_cost = {(sr, sc, 0, 1): 0}
    backtrack = {}
    best_cost = float("inf")
    end_states = set()
    
    while pq:
        cost, r, c, dr, dc = heapq.heappop(pq)
        
        if cost > lowest_cost.get((r, c, dr, dc), float("inf")):
            continue
            
        if grid[r][c] == END:
            if cost > best_cost:
                break
            best_cost = cost
            end_states.add((r, c, dr, dc))
            
        for new_cost, nr, nc, ndr, ndc in [
            (cost + 1, r + dr, c + dc, dr, dc),
            (cost + 1000, r, c, dc, -dr),
            (cost + 1000, r, c, -dc, dr)
        ]:
            if nr not in range(ROWS) or nc not in range(COLS) or grid[nr][nc] == WALL:
                continue
                
            lowest = lowest_cost.get((nr, nc, ndr, ndc), float("inf"))
            if new_cost > lowest:
                continue
                
            if new_cost < lowest:
                backtrack[(nr, nc, ndr, ndc)] = set()
                lowest_cost[(nr, nc, ndr, ndc)] = new_cost
                
            backtrack[(nr, nc, ndr, ndc)].add((r, c, dr, dc))
            heapq.heappush(pq, (new_cost, nr, nc, ndr, ndc))
    
    states = deque(end_states)
    seen = set(end_states)
    
    while states:
        key = states.popleft()
        for last in backtrack.get(key, []):
            if last in seen:
                continue
            seen.add(last)
            states.append(last)
    
    unique_tiles = {(r, c) for r, c, _, _ in seen}
    
    if PRINT:
        solution_grid = [row[:] for row in grid]
        for r, c in unique_tiles:
            if solution_grid[r][c] not in (START, END):
                solution_grid[r][c] = "O"
        for row in solution_grid:
            print("".join(row))
        print()
    
    return len(unique_tiles)

def parse():
    with open(filename) as f:
        inp = f.read()
    grid = list(map(list, inp.split("\n")))
    return grid

if __name__ == "__main__":

    filename = sys.argv[1] if len(sys.argv) > 1 else f"../data/{Path(__file__).stem}.txt"

    print("part1=" + str(part1()))
    print("part2=" + str(part2()))

File: cpp/day16.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <queue>
#include <set>
#include <map>
#include <tuple>
#include <algorithm>
#include <climits>

// Constants for grid elements
const char START = 'S';
const char EMPTY = '.';
const char WALL = '#';
const char END = 'E';

// Directions
enum Direction { NORTH = 0, EAST = 1, SOUTH = 2, WEST = 3 };
const int dr[4] = {-1, 0, 1, 0};
const int dc[4] = {0, 1, 0, -1};

// Parse the input file and return the grid
std::vector<std::vector<char>> parse(const std::string& filename) {
    std::ifstream infile(filename);
    std::string line;
    std::vector<std::vector<char>> grid;
    while (std::getline(infile, line)) {
        if (!line.empty()) {
            grid.push_back(std::vector<char>(line.begin(), line.end()));
        }
    }
    return grid;
}

// Part 1: Dijkstra-like search for shortest path from S to E
int part1(const std::vector<std::vector<char>>& grid) {
    int ROWS = grid.size(), COLS = grid[0].size();
    int sr = -1, sc = -1;
    for (int r = 0; r < ROWS; ++r) {
        for (int c = 0; c < COLS; ++c) {
            if (grid[r][c] == START) {
                sr = r; sc = c;
            }
        }
    }
    using State = std::tuple<int, int, int, int, std::vector<std::tuple<int, int, int>>>; // (score, r, c, dir, history)
    std::priority_queue<State, std::vector<State>, std::greater<State>> pq;
    pq.emplace(0, sr, sc, EAST, std::vector<std::tuple<int, int, int>>{});
    std::set<std::tuple<int, int, int, int>> visited;
    int final_score = -1;
    while (!pq.empty()) {
        auto [score, r, c, dir, history] = pq.top(); pq.pop();
        int drc = dr[dir], dcc = dc[dir];
        if (r < 0 || r >= ROWS || c < 0 || c >= COLS || grid[r][c] == WALL || visited.count({r, c, drc, dcc})) continue;
        if (grid[r][c] == END) {
            final_score = score;
            break;
        }
        visited.insert({r, c, drc, dcc});
        auto new_history = history;
        new_history.emplace_back(r, c, dir);
        // go straight
        pq.emplace(score + 1, r + drc, c + dcc, dir, new_history);
        // turn right
        pq.emplace(score + 1000, r, c, (dir + 1) % 4, new_history);
        // turn left
        pq.emplace(score + 1000, r, c, (dir + 3) % 4, new_history);
    }
    return final_score;
}

// Part 2: Dijkstra-like search, count unique tiles in all minimal-cost paths
int part2(const std::vector<std::vector<char>>& grid) {
    int ROWS = grid.size(), COLS = grid[0].size();
    int sr = -1, sc = -1;
    for (int r = 0; r < ROWS; ++r) {
        for (int c = 0; c < COLS; ++c) {
            if (grid[r][c] == START) {
                sr = r; sc = c;
                break;
            }
        }
        if (sr != -1) break;
    }
    using State = std::tuple<int, int, int, int, int>; // (cost, r, c, dr, dc)
    auto cmp = [](const State& a, const State& b) { return std::get<0>(a) > std::get<0>(b); };
    std::priority_queue<State, std::vector<State>, decltype(cmp)> pq(cmp);
    pq.emplace(0, sr, sc, 0, 1); // start facing east
    std::map<std::tuple<int, int, int, int>, int> lowest_cost;
    std::map<std::tuple<int, int, int, int>, std::set<std::tuple<int, int, int, int>>> backtrack;
    int best_cost = INT_MAX;
    std::set<std::tuple<int, int, int, int>> end_states;
    lowest_cost[{sr, sc, 0, 1}] = 0;
    while (!pq.empty()) {
        auto [cost, r, c, drc, dcc] = pq.top(); pq.pop();
        if (cost > lowest_cost[{r, c, drc, dcc}]) continue;
        if (grid[r][c] == END) {
            if (cost > best_cost) break;
            best_cost = cost;
            end_states.insert({r, c, drc, dcc});
        }
        // go straight
        int nr = r + drc, nc = c + dcc;
        if (nr >= 0 && nr < ROWS && nc >= 0 && nc < COLS && grid[nr][nc] != WALL) {
            int new_cost = cost + 1;
            auto key = std::make_tuple(nr, nc, drc, dcc);
            if (!lowest_cost.count(key) || new_cost < lowest_cost[key]) {
                backtrack[key] = {};
                lowest_cost[key] = new_cost;
            }
            if (new_cost <= lowest_cost[key]) {
                backtrack[key].insert({r, c, drc, dcc});
                pq.emplace(new_cost, nr, nc, drc, dcc);
            }
        }
        // turn right
        int ndr = dcc, ndc = -drc;
        if (grid[r][c] != WALL) {
            int new_cost = cost + 1000;
            auto key = std::make_tuple(r, c, ndr, ndc);
            if (!lowest_cost.count(key) || new_cost < lowest_cost[key]) {
                backtrack[key] = {};
                lowest_cost[key] = new_cost;
            }
            if (new_cost <= lowest_cost[key]) {
                backtrack[key].insert({r, c, drc, dcc});
                pq.emplace(new_cost, r, c, ndr, ndc);
            }
        }
        // turn left
        ndr = -dcc; ndc = drc;
        if (grid[r][c] != WALL) {
            int new_cost = cost + 1000;
            auto key = std::make_tuple(r, c, ndr, ndc);
            if (!lowest_cost.count(key) || new_cost < lowest_cost[key]) {
                backtrack[key] = {};
                lowest_cost[key] = new_cost;
            }
            if (new_cost <= lowest_cost[key]) {
                backtrack[key].insert({r, c, drc, dcc});
                pq.emplace(new_cost, r, c, ndr, ndc);
            }
        }
    }
    // Backtrack to find all unique tiles in minimal-cost paths
    std::deque<std::tuple<int, int, int, int>> states(end_states.begin(), end_states.end());
    std::set<std::tuple<int, int, int, int>> seen(end_states.begin(), end_states.end());
    while (!states.empty()) {
        auto key = states.front(); states.pop_front();
        for (const auto& last : backtrack[key]) {
            if (seen.count(last)) continue;
            seen.insert(last);
            states.push_back(last);
        }
    }
    std::set<std::pair<int, int>> unique_tiles;
    for (const auto& s : seen) {
        int r, c, drc, dcc;
        std::tie(r, c, drc, dcc) = s;
        unique_tiles.insert({r, c});
    }
    return unique_tiles.size();
}

int main(int argc, char* argv[]) {
    std::string filename;
    if (argc > 1) {
        filename = argv[1];
    } else {
        filename = "../data/day16.txt";
    }
    std::vector<std::vector<char>> grid = parse(filename);
    std::cout << "part1=" << part1(grid) << std::endl;
    std::cout << "part2=" << part2(grid) << std::endl;
    return 0;
} 